text/0000-default-and-expanded-rustc-errors.md

```diff
@@ -0,0 +1,290 @@
+* Prints messages about lines often out of order.  This makes it difficult for the developer to glance at the error and recognize why the error is occuring
+* Lacks a clear visual break between errors.  As more errors occur it becomes more difficult to tell them apart.
+* Uses technical terminology that is difficult for new users who may be unfamiliar with compiler terminology or terminology specific to Rust.
+
+This RFC details a redesign of errors to focus more on the source the programmer wrote.  This format addresses the above concerns by eliminating clutter, following a more natural order for help messages, and pointing the user to both "what" the error is and "why" the error is occurring by using color-coded labels.  Below you can see the same error again, this time using the proposed format:
+
+[as text]
+```
+error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
+  --> src/test/compile-fail/borrowck/borrowck-borrow-from-owned-ptr.rs:29:22
+   |
+28 |      let bar1 = &mut foo.bar1;
+   |                      -------- first mutable borrow occurs here
+29 |      let _bar2 = &mut foo.bar1;
+   |                       ^^^^^^^^ second mutable borrow occurs here
+30 |      *bar1;
+31 |  }
+   |  - first borrow ends here
+```
+
+[as image]
+
+<a href="http://www.jonathanturner.org/images/new_errors_3.png"><img src="http://www.jonathanturner.org/images/new_errors_3.png" height="200" width="750" ></a>
+
+*Example of the same borrow check error in the proposed format*
+
+## Expanded error format (revised --explain)
+
+Languages like Elm have shown how effective an educational tool error messages can be if the explanations like our --explain text are mixed with the user's code.  As mentioned earlier, it's crucial for Rust to be easy-to-use, especially since it introduces a fair number of concepts that may be unfamiliar to the user.  Even experienced users may need to use --explain text from time to time when they encounter unfamiliar messages.
+
+While we have --explain text today, it uses generic examples that require the user to mentally translate the given example into what works for their specific situation.
+
+```
+You tried to move out of a value which was borrowed. Erroneous code example:
+
+use std::cell::RefCell;
+
+struct TheDarkKnight;
+
+impl TheDarkKnight {
+    fn nothing_is_true(self) {}
+}
+...
+```
+
+*Example of the current --explain (showing E0507)*
+
+To help users, this RFC proposes a new `--explain errors`.  This new mode is more textual error reporting mode that gives additional explanation to help better understand compiler messages.  The end result is a richer, on-demand error reporting style.
+
+```
+error: cannot move out of borrowed content
+   --> /Users/jturner/Source/errors/borrowck-move-out-of-vec-tail.rs:30:17
+
+I’m trying to track the ownership of the contents of `tail`, which is borrowed, through this match statement:
```

Comment:
Thanks for the feedback.  I should have probably reiterated again in this section that this RFC's focus is the error format, not the error writing style.  The way these errors are written will be a follow-up RFC as we get more experience with what works best.

I list some evidence, by way of mentioning it being praised by noted programmers, but anecdotally, I've been watching the feedback from other programmers around the web.  They praise the elm style for being a) one of the best error message styles, b) so thorough they use it as a way to pick back up an editing session if it's cut short on a train ride, c) more new user-friendly than other error styles.

Note, too, that we're not saying that the expanded style is better for all users, but rather it's one we should support for users who wish to have more information.  Some users may enable `--explain errors` and leave it on for all projects, and that would be fine, while others may prefer more compact formats.